### PR TITLE
Register json serde types for reflection

### LINKED ISSUE #18276: Kafka message payload types are not registered for reflection
## Describe the bug
Serde message payloads need to be registered for reflection in native builds.
If these classes were not registered by another means (i.e resteasy extension) message serde does not work in native mode.

### Expected behavior

Message serde (jackson/jsonb) works in native mode.

### Actual behavior

For example with Jackson serialization fails with a similar stacktrace:

```
Caused by: java.lang.RuntimeException: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class org.acme.kafka.model.Quote and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS)
	at io.quarkus.kafka.client.serialization.ObjectMapperSerializer.serialize(ObjectMapperSerializer.java:37)
	at org.apache.kafka.common.serialization.Serializer.serialize(Serializer.java:62)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:925)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:885)
	at io.smallrye.reactive.messaging.kafka.impl.ReactiveKafkaProducer.lambda$send$3(ReactiveKafkaProducer.java:96)
	at io.smallrye.context.impl.wrappers.SlowContextualConsumer.accept(SlowContextualConsumer.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:22)
	... 7 more
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class org.acme.kafka.model.Quote and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS)
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1276)
	at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:400)
	at com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.failForEmpty(UnknownSerializer.java:71)
	at com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.serialize(UnknownSerializer.java:33)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4487)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3701)
	at io.quarkus.kafka.client.serialization.ObjectMapperSerializer.serialize(ObjectMapperSerializer.java:34)
	... 13 more
``` 

## To Reproduce

1. Setup app with `smallrye-reactive-kafka` extension
2. Send/receive serde'd object to/from Kafka
3. Make sure the object is not already registered for reflection by another extension, (i.e REST endpoint payload)
4. Run the app in native mode

### PR DESCRIPTION
**Pull request by ozangunalp**
Fixes #18276

### CODE REVIEW

**Comment by gsmet** [comment]
@Ladicek it would be nice if you could have a look. I would like to release 2.0.2 tomorrow and it looks like something we would want

**Comment by Ladicek** [comment]
I was on PTO, looking at this now.

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java [hunk 1]
```diff
@@ -532,4 +548,113 @@ private String serializerDeserializerFor(DefaultSerdeDiscoveryState discovery, T
...
+        return getSubclassOfWithTypeArgument(index, DotNames.OBJECT_MAPPER_SERIALIZER, serializerName) ||
+                getSubclassOfWithTypeArgument(index, DotNames.JSONB_SERIALIZER, serializerName);
+    }
+
+    private boolean isSubclassOfJsonDeserializer(IndexView index, DotName serializerName) {
+        return getSubclassOfWithTypeArgument(index, DotNames.OBJECT_MAPPER_DESERIALIZER, serializerName) ||
+                getSubclassOfWithTypeArgument(index, DotNames.JSONB_DESERIALIZER, serializerName);
+    }
+
+    private boolean getSubclassOfWithTypeArgument(IndexView index, DotName superclass, DotName expectedType) {
```
Comment (Ladicek) [inline extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java:652]: The name here is copied from a method that does something different I think. Rename to `isSubclassOf`?

Comment (ozangunalp) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java:652]: I agree!

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/ReflectiveClassForValueSerializerPaylaodTest.java [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+import io.quarkus.smallrye.reactivemessaging.kafka.deployment.DefaultSerdeConfigTest.JsonbDto;
+import io.quarkus.smallrye.reactivemessaging.kafka.deployment.DefaultSerdeConfigTest.JsonbDtoSerializer;
+import io.smallrye.config.SmallRyeConfigBuilder;
+import io.smallrye.config.common.MapBackedConfigSource;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.reactive.messaging.MutinyEmitter;
+import io.smallrye.reactive.messaging.kafka.KafkaRecord;
+import io.smallrye.reactive.messaging.kafka.Record;
+
+class ReflectiveClassForValueSerializerPaylaodTest {
```
Comment (Ladicek) [inline extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/ReflectiveClassForValueSerializerPaylaodTest.java:40]:
```suggestion
class ReflectiveClassForValueSerializerPayloadTest {
```

:-)

Comment (ozangunalp) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/ReflectiveClassForValueSerializerPaylaodTest.java:40]: It took me 3 minutes and numerous ctrl-f to find that out :)

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java [hunk 3]
```diff
@@ -532,4 +548,113 @@ private String serializerDeserializerFor(DefaultSerdeDiscoveryState discovery, T
...
+
+        processIncomingForReflectiveClassPayload(index, config,
+                (annotation, payloadType) -> produceReflectiveClass(reflectiveClass, annotation, payloadType));
+
+        processIncomingChannelForReflectiveClassPayload(index, config,
+                (annotation, payloadType) -> produceReflectiveClass(reflectiveClass, annotation, payloadType));
+
+    }
+
+    void produceReflectiveClass(BuildProducer<ReflectiveClassBuildItem> reflectiveClass, AnnotationInstance annotation,
```
Comment (Ladicek) [inline extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java:574]: The `annotation` parameter is useless?

**Review by Ladicek — APPROVED** [review]
With the refactoring of the serde autodetection code, I think I would get rid of a lot of methods that now just call another method, but that's fine and can be done later. I also added a few small comments that can be done later as well. Otherwise LGTM.

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/ReflectiveClassForValueSerializerPaylaodTest.java (hunk 2 above)
Comment (Ladicek) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/ReflectiveClassForValueSerializerPaylaodTest.java:40]: I'm blind to my own typos, but can spot others' really quickly :rofl:

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java (hunk 3 above)
Comment (ozangunalp) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java:574]: I needed it while testing but then changed how to call in tests. I'll remove it.

**Comment by ozangunalp** [comment]
@Ladicek I agree with you on the number of methods. I tested some inlines but in the end, found that having separate methods was more readable.

**Comment by Ladicek** [comment]
Fair enough. When I get to #18195, I'll probably have to do another refactoring, so we'll see :-)

**Review by Ladicek — APPROVED** [review]

**Comment by Ladicek** [comment]
Wanna squash the commits in some way, maybe?

**Comment by gsmet** [comment]
I rebased and squashed the last 2 commits. Let's see how it goes.

**Comment by ozangunalp** [comment]
@gsmet Thanks! I didn't saw @Ladicek 's second comment.

**Pull request merged**